util/apicaps/caps.go

```diff
@@ -0,0 +1,161 @@
+package apicaps
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	pb "github.com/moby/buildkit/util/apicaps/pb"
+	"github.com/pkg/errors"
+)
+
+type PBCap = pb.APICap
+
+// ExportedProduct is the name of the product using this package.
+// Users vendoring this library may override it to provide better versioning hints
+// for their users (or set it with a flag to buildkitd).
+var ExportedProduct string = "buildkit"
+
+// CapStatus defines the stability properties of a capability
+type CapStatus int
+
+const (
+	// CapStatusStable should never be changed in backwards incompatible manner unless
+	// there is a serious security issue.
+	CapStatusStable CapStatus = iota
+	// CapStatusExperimental may be removed in the future. If incompatible changes are
+	// made the previous ID is disabled and new is added.
+	CapStatusExperimental
+	// CapStatusPrerelease is same as CapStatusExperimental that can be used for new
+	// features before they move to stable.
+	CapStatusPrerelease
+)
+
+// CapID is type for capability identifier
+type CapID string
+
+// Cap describes an API feature
+type Cap struct {
+	ID                  CapID
+	Name                string // readable name, may contain spaces but keep in one sentence
+	Status              CapStatus
+	Enabled             bool
+	Deprecated          bool
+	SupportedHint       map[string]string
```

Comment:
Would be useful to add a helper function in the future to allow downstream to set supported hints for their product.
In which case I wonder if ExportedProduct could not be moved to a more buildkit-specific package.

Reply:
Actually it's fine to keep ExportedProduct, it's a way for downstream to register itself.